### PR TITLE
Add support for Raspberry Pi Pico (RP2040) to community-2021

### PR DESCRIPTION
**Pull request by damaki**
This is resubmission of #51, rebased and updated to the `community-2021` branch, with a new boot2 written in Ada which replaces the previous `boot2__w25q080.S` (now deleted) that was copied from the pico-sdk.

The main addition since #51 is the addition of the new stage 2 bootloader, located in the `arm/rpi/rp2040/boot2/` directory. The code is mostly Ada code, with a little bit of assembly to handle the LR register to vector into the user application. Two flash devices are supported:
* generic_03 - this is compatible with any flash device that supports the 03h read command. The trade-off is that it uses standard SPI mode, so is slower than using QSPI.
* w25q080 - this is the flash device used for the Pico board, and operates in QSPI mode.

The boot2 uses an out-of-band build process which produces a .S with the boot2 binary that is then included in the runtime build. The boot2 binaries have been pre-built from the Ada sources and are stored in the `arm/rpi/rp2040/boot2/generated/` directory. We did it this way as the binaries need to be padded to 256 bytes with the last four bytes containing a CRC32, and we couldn't find any easy way  to do the padding and CRC32 within the bb-runtimes build system. The scripts to generate these .S files are also in the `boot2` directory.

Many thanks to @JeremyGrosser for rebasing the changes to the 2021 branch and for helping with the boot2 work.

### CODE REVIEW

File: install.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ def usage():
                         'nrf52840', 'cortex-m0', 'cortex-m0p', 'cortex-m1',
                         'cortex-m3', 'cortex-m4', 'cortex-m4f', 'cortex-m7f',
                         'cortex-m7df', 'cortex-m23', 'cortex-m33f',
-                        'cortex-m33df'
+                        'cortex-m33df', 'rpi-pico',
```
Comment (Fabien-Chouteau) [inline install.py:33]: You can add the `rpi-pico-smp` here.

File: arm/cortexm.py [hunk 2]
```diff
@@ -1140,6 +1140,179 @@ def compiler_switches(self):
         return ('-mlittle-endian', '-mthumb', '-mfloat-abi=soft',
                 '-mcpu=cortex-m0plus')
 
+    @property
+    def system_ads(self):
```
Comment (Fabien-Chouteau) [inline arm/cortexm.py:1144]: This is supposed to be inherited from `CortexM0`. Any specific reason to re-define it?

File: arm/cortexm.py [hunk 3]
```diff
@@ -1140,6 +1140,179 @@ def compiler_switches(self):
...
+        return {'zfp': 'system-xi-arm.ads'}
+
+
+class RP2040(CortexM0P):
```
Comment (Fabien-Chouteau) [inline arm/cortexm.py:1148]: I would prefer to see this class after all the CorexM* classes.

**Review by Fabien-Chouteau — CHANGES_REQUESTED** [review]
Awesome stuff @damaki and @JeremyGrosser

Two general comments on top of the specify ones:
 - If you can squash all commits in one that would be better
 - All Ada source files that are not generated (from SVD) must have the GNAT RUN-TIME COMPONENTS header:
```ada
------------------------------------------------------------------------------
--                                                                          --
--                         GNAT RUN-TIME COMPONENTS                         --
--                                                                          --
--                      Copyright (C) 2021, AdaCore                         --
--                                                                          --
-- GNAT is free software;  you can  redistribute it  and/or modify it under --
-- terms of the  GNU General Public License as published  by the Free Soft- --
-- ware  Foundation;  either version 3,  or (at your option) any later ver- --
-- sion.  GNAT is distributed in the hope that it will be useful, but WITH- --
-- OUT ANY WARRANTY;  without even the  implied warranty of MERCHANTABILITY --
-- or FITNESS FOR A PARTICULAR PURPOSE.                                     --
--                                                                          --
-- As a special exception under Section 7 of GPL version 3, you are granted --
-- additional permissions described in the GCC Runtime Library Exception,   --
-- version 3.1, as published by the Free Software Foundation.               --
--                                                                          --
-- You should have received a copy of the GNU General Public License and    --
-- a copy of the GCC Runtime Library Exception along with this program;     --
-- see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see    --
-- <http://www.gnu.org/licenses/>.                                          --
--                                                                          --
-- GNAT was originally developed  by the GNAT team at  New York University. --
-- Extensive contributions were provided by Ada Core Technologies Inc.      --
--                                                                          --
------------------------------------------------------------------------------
```

**Comment by damaki** [comment]
> * All Ada source files that are not generated (from SVD) must have the GNAT RUN-TIME COMPONENTS header:

@JeremyGrosser you were the original author of `s-bootro.ads` and `s-bootro.adb` (with some modifications from myself) and these files have your copyright notice on it and the BSD-3-Clause license. Are you happy for me to apply the "GNAT RUN-TIME COMPONENTS" header to these files? It would require these files to be relicensed from BSD-3-Clause to GPLv3 with the runtime exception. I.e. the copyright header would look like this:

```
------------------------------------------------------------------------------
--                                                                          --
--                         GNAT RUN-TIME COMPONENTS                         --
--                                                                          --
--                  Copyright (C) 2021, Jeremy Grosser                      --
--                      Copyright (C) 2021, AdaCore                         --
--                                                                          --
-- GNAT is free software;  you can  redistribute it  and/or modify it under --
-- terms of the  GNU General Public License as published  by the Free Soft- --
-- ware  Foundation;  either version 3,  or (at your option) any later ver- --
-- sion.  GNAT is distributed in the hope that it will be useful, but WITH- --
-- OUT ANY WARRANTY;  without even the  implied warranty of MERCHANTABILITY --
-- or FITNESS FOR A PARTICULAR PURPOSE.                                     --
--                                                                          --
-- As a special exception under Section 7 of GPL version 3, you are granted --
-- additional permissions described in the GCC Runtime Library Exception,   --
-- version 3.1, as published by the Free Software Foundation.               --
--                                                                          --
-- You should have received a copy of the GNU General Public License and    --
-- a copy of the GCC Runtime Library Exception along with this program;     --
-- see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see    --
-- <http://www.gnu.org/licenses/>.                                          --
--                                                                          --
-- GNAT was originally developed  by the GNAT team at  New York University. --
-- Extensive contributions were provided by Ada Core Technologies Inc.      --
--                                                                          --
------------------------------------------------------------------------------
```

Likewise, you wrote the original code for `boot2.adb` and `boot2_generic_03.ads` which lacked any copyright header. Are you happy for me to put the copyright/license notice on those files too?

**Comment by JeremyGrosser** [comment]
Yes, you can replace the copyright header on all the parts I wrote.

**Comment by damaki** [comment]
Thanks @JeremyGrosser .

@Fabien-Chouteau I've made the changes you requested, and have squashed everything to a single commit (f183e55).

**Comment by damaki** [comment]
FYI I've just found an issue where building a newly-created project (with the default empty main subprogram) using the `ravenscar-sfp-rpi-pico-smp` runtime fails to link with the following error:

```
d:/gnat/2021-arm-elf/bin/../lib/gcc/arm-eabi/10.3.1/../../../../arm-eabi/bin/ld.exe: D:\gnat\2021-arm-elf\arm-eabi\lib\gnat\ravenscar-sfp-rpi-pico-smp\adalib\libgnat.a(start-rom.o): in function `__vectors':
(.vectors+0x7c): undefined reference to `__gnat_poke_handler'
d:/gnat/2021-arm-elf/bin/../lib/gcc/arm-eabi/10.3.1/../../../../arm-eabi/bin/ld.exe: (.vectors+0x80): undefined reference to `__gnat_poke_handler'
collect2.exe: error: ld returned 1 exit status
```

The main subprogram looks like this:
```Ada
procedure Main is
begin
   null;
end Main;
```

This only occurs with the SFP runtime for the SMP variant. The `ravenscar-full-rpi-pico-smp` and non-SMP runtimes build successfully with an empty program.

If I add `with Ada.Real_Time;` to the `main.adb` then the link error goes away, but I'd like to find a way to resolve this as it gives a pretty bad impression if a newly created project fails to build.

**Comment by Fabien-Chouteau** [comment]
This is because all the run-time tasking code is only linked when tasking is actually used.
`__gnat_poke_handler` should be declared weak with a replacement function (similar to `__gnat_irq_trap`).

**Comment by damaki** [comment]
That's exactly the solution I'm implementing and testing right now :)

**Comment by damaki** [comment]
FYI I've just pushed another fix to resolve a problem with the SMP runtime where resetting core0 via a debugger would cause it to get stuck waiting for core1 to start-up. I added a `Reset_Core1` procedure to `s-bbsumu__rp2040.adb` to ensure that core1 is reset to a known state when core0 starts up.

**Comment by damaki** [comment]
I've pushed some more changes which add support for a bunch of other boards using the RP2040, including boards from Arduino, Adafruit, Pimoroni, and Sparkfun.

The main improvement was to to update the boot2 implementation to better support the various flash chips used on the supported boards (see `arm/rpi/rp2040/boot2/README` for details). The other change was to make the XOSC startup delay configurable to ensure reliable startup, particularly for Adafruit boards which have a slower startup time.

As before, I've squashed everything to a single commit.

**Comment by Fabien-Chouteau** [comment]
Thank you @damaki and @JeremyGrosser.
Because of internal process we did not use GitHub to merge the changes, so the PR is not marked as complete.
But this changes are in: https://github.com/AdaCore/bb-runtimes/commit/28eeb375ffa9f8f53011154365736383f780a860

Thanks for this awesome contribution.

**Pull request closed without merging**